PlayFabSDK/source/Qos/PlayFabQosApi.cs

```diff
@@ -0,0 +1,163 @@
+            PlayFabResult<ListQosServersResponse> response = await PlayFabMultiplayerAPI.ListQosServersAsync(request);
+            
+            if (response?.Error != null)
+            {
+                LogQos("Could not get the server list from thunderhead\n Error : " + response.Error.ErrorMessage);
+            }
+
+            foreach (QosServer qosServer in response.Result.QosServers)
+            {
+                if (!qosServer.Region.HasValue)
+                {
+                    continue;
+                }
+
+                _dataCenterMap[qosServer.Region.Value] = qosServer.ServerUrl;
+            }
+        }
+
+        private async Task<List<QosRegionResult>> GetSortedRegionLatencies(int timeoutMs)
+        {
+            List<Task<QosRegionResult>> asyncPingResults = _dataCenterMap.Select(async kvp =>
+            {
+                var regionPinger = new RegionPinger(kvp.Value, kvp.Key);
+                return await regionPinger.PingAsync(timeoutMs);
+            }).ToList();
+
+            await Task.WhenAll(asyncPingResults);
+            return asyncPingResults.Select(t => t.Result).OrderBy(t => t.LatencyMs).ToList();
+        }
+
+        private async Task<bool> SendResultsToPlayFab(QosResult result)
+        {
+            var eventContents = new EventContents
+            {
+                Name = "qos_result",
+                EventNamespace = "playfab.servers",
+                Payload = result
+            };
+
+            var writeEventsRequest = new WriteEventsRequest
+            {
+                Events = new List<EventContents> {eventContents}
+            };
+
+            PlayFabResult<WriteEventsResponse> response = null;
+            try
+            {
+                response = await WriteTelemetryAsync(writeEventsRequest);
+            }
+            catch (Exception exception)
+            {
+                LogQos(exception.ToString());
+                return false;
+            }
+
+            if (response == null || response.Error != null)
+            {
+                LogQos(response?.Error.ErrorMessage);
+                return false;
+            }
+
+            return true;
+        }
+
+        private static async Task<PlayFabResult<WriteEventsResponse>> WriteTelemetryAsync(WriteEventsRequest request, object customData = null, Dictionary<string, string> extraHeaders = null)
+        {
+            if ((request?.AuthenticationContext?.EntityToken ?? PlayFabSettings.staticPlayer.EntityToken) == null) throw new PlayFabException(PlayFabExceptionCode.EntityTokenNotSet, "Must call GetEntityToken before calling this method");
+
+            var httpResult = await PlayFabHttp.DoPost("/Event/WriteTelemetryEvents", request, "X-EntityToken", PlayFabSettings.staticPlayer.EntityToken, extraHeaders);
+            if (httpResult is PlayFabError)
+            {
+                var error = (PlayFabError)httpResult;
+                PlayFabSettings.GlobalErrorHandler?.Invoke(error);
```

Comment:
GlobalErrorHandler existed when the only SDK feature was the API.
And, I'm not certain I want to continue the pattern as-is.
I think the QOS system could have its own ErrorHandler callback, but I think we should rename/phase-out the GlobalErrorHandler in general.  (The error type itself is very specific to the PlayFab API, and it's only a coincidence that it matches your error here).

Reply:
GlobalErrorHandler is also not compatible with the general move towards instanced APIs and such.

Reply:
Here I just create a simple call to WriteTelemetryEvents, since its API doesn't exist in SDK, and there is no plan to create that API in SDK. It is the same as other direct generated APIs. So I would want to keep it consistent with WriteEvents API call.